credentials/tls/certprovider/distributor.go

```diff
@@ -0,0 +1,108 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package certprovider
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/internal/grpcsync"
+)
+
+// Distributor makes it easy for provider implementations to furnish new key
+// materials by handling synchronization between the producer and consumers of
+// the key material.
+//
+// Provider implementations must embed this type into themselves and:
+// - Whenever they have new key material, they should invoke the Set() method.
+// - When users of the provider call KeyMaterial(), it will be handled by the
+//   distributor which will return the most up-to-date key material furnished
+//   by the provider.
+// - When users of the provider call Close(), the channel returned by the
+//   Done() method will be closed. So, provider implementations can select on
```

Comment:
In what situation would provider wait on `Done()`?
Can they all be handled by overriding `Close()` (so provider creates it's own done channel)?

Reply:
If a `Provider` implementation is embedding a `Distributor` and is not implementing `Close`, then they must wait on the `Done()` channel to perform any cleanup. There is no other way for them to be notified that they have been closed.

Providers which implement `Close` and also use a `Distributor`, they should invoke `Distributor.Close` from their `Close`.

We don't want the `Distributor` to depend on something on the `Provider`. I don't see how a `Done` channel on the `Provider` would be useful.

Reply:
This has now been changed to expose a `Ctx` field that the `Provider` implementations can select on.